### ISSUE TITLE
Scanner: Changed artist names remain in database

### ISSUE THREAD

**Issue by frank1969b** [LMS-Community/slimserver#704]
As discussed in the forum, here's the bug I find most annoying:

If I correct/change an (track) artist name (eg for it has been misspelled or there are more artists of the same name that I have to split like "Pauline (02)", "Pauline (03)" the OLD artist name is (after a changed&new scan) not deleted from the database.

Example:
I have tracks from different artists named "Pauline" (one from Germany, one from France, one from Finland...).
So I split them to "Pauline", "Pauline (02)", "Pauline (03)"....
After a "changed&new scan" browsing for "Pauline" ALL tracks/albums are still shown (also those now credited to "Pauline (02)", "Pauline (03)"...).
If I click on the album cover / entry, it only shows the album cover and information and the word "empty" (instead of any track information).

The other way works correctly:
If I browse for eg "Pauline (02)", only tracks credited to "Pauline (02)" are shown.
If I click on the tracks/entrys, only "Pauline (02)" is mentioned as artist (though it is only found browing for Pauline)".

I see the same behaviour if I correct spelling like changing "Armin Van Buuren" => "Armin van Buuren" - after a new&changed there still remain 2 artists in my database "Armin Van Buuren" and "Armin van Buuren", though if I click "Armin Van Buuren" all browsing ends in "empty" entries.

Or another thing:
I have a track credited to "Armin van Buuren & Timmy Trumpet", that I change to "Armin van Buuren\\Timmy Trumpet" - it afterwards is in the database under 3 entries "Armin van Buuren", "Timmy Trumpet" (both wanted behaviour) and "Armin van Buuren + Timmy Trumpet" with an empty entry...

Would be GREAT if this could be fixed :)

**Comment by mherger** [comment]
True... that bug is so old I would have bet your report was a duplicate. But I didn't find anything matching. And in your case a wipe & rescan, the usual workaround, isn't the way to go 😁.

**Comment by frank1969b** [comment]
Indeed, I see You know my "problem" ;)
But I guess with all of our music collections growing more (and I think on "long hand" only those with large collections will resist the tempation of "streaming only") it's not a good idea to do a wipe&rescan everyday (though for Your information after buying a new PC it is done in about 8-10 hours - but it means 8-10 hours without music a day  ;-) )

**Comment by mherger** [comment]
@frank1969b would it be possible to have a copy of your library.db when it's in such a broken state?

We are cleaning up contributors which, after renaming them, no longer exist. In that case the contributor and all album relations would be deleted. BUT if the contributor continues to exist, as you only renamed two out of the three, then I believe we don't remove the contributor -> album mapping for the other two contributors.

Maybe we'll need to verify whether tracks for the `$contrib && $orig->{album}` still exists, or clean up `contributor_albums` around here:

https://github.com/Logitech/slimserver/blob/f98f7466717e30e3ec0683443bdb6ece11d3ae17/Slim/Utils/Scanner/Local.pm#L1001-L1005

**Comment by mherger** [comment]
@frank1969b I added another sanity check to the scanner which would hopefully fix this issue. It's in 8.3 only. Would you please give it a try with the next build? Thanks!

**Comment by frank1969b** [comment]
@mherger , many thanks You care about the issue. Tested yesterday with 8.3 r1643414649 but saw no changes. Just picked me ONE example and do some tests and db views, will report later...

**Comment by michaelherger** [comment]
How did you test? Please note that the change affects the scan of newly changed files only. You'll have to change the files again for them to be rescanned. Rescanning without changing the affected files will not have any impact.

**Comment by frank1969b** [comment]
@mherger - yes, I'm aware. 
I did the following: I picked an artist that is misspelled in my list with 2 entries only (eg. "Afrojack & David Guetta" - without a splitter) and changed it on BOTH tracks to "Afrojack\\David Guetta" and did a rescan. The changed files ARE recognized, for after a rescan, they give (as artist) Afrojack, David Guetta (as two seperate entries).
So far so good.
BUT
I I search for "Afrojack" it still shows "Afrojack" and "Afrojack & David Guetta" as 2 artists. 
If I click the "Afrojack & David Guetta" entry with the 2 albums the track is on.
If I click on an album, it shows the cover and "empty".

So I took a look in the db.
- Table "contributors" still has "Afrojack & David Guetta" with # 256342
- Searching "contributor_album" for contributor #256342 it shows them on those 2 albums, both with role "1"
- Searching "contributor_track" for contributor #256342 gives NO results
- Searching "library_contributor" for contributor #256342 gives NO results
- Searching "albums" for those 2 albums gives (for both) the "Various Artists"-Artist (#142239) as "contributor". "compilation" ist set to "1" on both albums
- Searching "tracks" for the both tracks gives "157270" (which is "Afrojack") als "primary_artist"
- Searching "contributor_track" for track id gives 3 results, the "Various Artist"-ID with role 5 and the "Afrojack" and "David Guetta" ID with role 1 (but NOT the "misspelled artist"
- Searching "contributor_album" for contributor gives  1 "role 5"-result (the "Various Artists"-ID) and many "role 1"-result including the "Afrojack"-ID, the "David Guetta"-ID AND the "Afrojack & David Guetta"-ID (see above).

So this is all I can see...

If it matters:
In my options I have
"Browse Artists" set to "Use single, configurable list of artists"
"Compilations" set to "List compiltaion albums under each artist" and "Treat TPE2 MP3 tag as "ALBUM Artist"

Do I have any chance to see anything that matters to You in the log files (then I would pick another misspelled artist and try the same thing with maybe changed logging options)
At the moment
- "scan", "scan.import" and "scan.scanner" are set to "Error"
- "scan.auto" is set to "Debug"

Many thanks again
Frank

**Comment by michaelherger** [comment]
I probably only fixed the Pauline issue, but not the other one. I'll have to double check the searching. But can you confirm that by browsing to the artist using the regular menus you wouldn't see the empty albums any more?

**Comment by michaelherger** [comment]
Thank you very much for the db analysis. Unfortunately it shows that something in my change either did not get called, or doesn't work as expected: when a track changes, the scanner should check whether for any of the old artists ("Afrojack & David Guetta") there was still a track. If none was left, the contributor should be removed (which is not the case here). The newly added code should delete the entry from `contributor_album` if the contributor was still required for some other album, track or role. But that doesn't happen according to your analysis. Unfortunately the database logging would be crazy verbose for a scan with a library that size. It would be hard to find anything in there :-(. 

I'll have to play with your case again. As I said I concentrated more on a "Pauline" like case, which is slightly different. And who knows whether the compilation doesn't play a role, too... So I'll have to grab a compilation with two artists in one name, and split them.

**Comment by michaelherger** [comment]
Hmm.... tried that: I took a compilation with one track by artist "you & me". Did a wipe & rescan. That literal artist is found in the contributor table. I then changed the artist to "you;me" (I have the semicolon as separated for artists). After a simple rescan the previous artist was gone from the contributor table, and two new were added. 

In short: I can't reproduce your issue. Maybe I don't fully understand the exact case.

**Comment by frank1969b** [comment]
@mherger , of course I'm testing as soon as there is a new nightl (just testing the r1643570786 - though I guess there is no fix yet?)
Hm, the "you & me" situation is EXACTLY my example. The only difference is, I change it to "you\\me" in "mp3tag" resp. to "you; me" if I do it in other tools like "mediamonkey" (what leads to "you\\me" as well..)
But THIS example I did with the "\\" option (without a space...)

**Comment by frank1969b** [comment]
I will have another try with a non-compilation album...

**Comment by frank1969b** [comment]
@mherger I did some more testing and got a few results:
- It doesn't seem to have any impact, if there is a compilation or a single artist album
- I tried Your "You&me" example in 2 ways and saw 2 different behaviours depending if "you" and "me" have been in the database before:
1. I did the same as you, created a track from a "You&me" artist, rescanned, it was in the database. Then I splitted into "you\\me", "simple" re-scanned => "you&me" was gone from database, "you" was in database, "me" was in database => FINE!
2. I created one track from a "you&me" artist, one track from a "you" artist, one track from the "me" artist. I scanned - all 3 acts were in the database.
Then I once again splitted into "you\\me", "simple" rescanned... and now the "you" artist had 2 tracks, the "me" artist had 2 tracks - and the "you&me" was still there with an empty entry (as I desribed before).

So it maybe makes a difference, if "you" and "me" are new artists or if they have been in the database before. Maybe this may ring a bell for You?

Another strange behaviour I watched in "2.".

If I search for "You&me" it is still there in database (as I described).
If I click on the "you&me"-artist, the album cover is shown AND the title is shown.
If I click on the track, artists as shown as "you" and "me" (NOT "you&me") as expected.

Now I search again for "you&me", it is still there (as it was before).
If I NOW click on the "you&me"-Artist, the album cover is shown ... but no title, only the word "empty"...

Just to be sure:
I am sure, that scanner scanned and changed the files (I controlled the changedate and it is obvious for it changed the "you" and "me" artists...)
... and I'm aware, once the issue is fixed, I have to do a clear scan to my database ;-)

May this help?
